src/prettyPrinter.ts

```diff
@@ -1,804 +1,889 @@
+//
+
+export const unaryOperatorType: Record<A.AstUnaryOperation, "post" | "pre"> = {
+    "+": "pre",
+    "-": "pre",
+    "!": "pre",
+    "~": "pre",
+
+    "!!": "post",
+};
+
+export const checkPostfix = (operator: A.AstUnaryOperation) =>
+    unaryOperatorType[operator] === "post";
 
 /**
- * Provides methods to format and indent Tact code.
+ * Description of precedence of certain type of AST node
  */
-export class PrettyPrinter {
+export type Precedence = {
     /**
-     * @param indentLevel Initial level of indentation.
-     * @param indentSpaces Number of spaces per indentation level.
+     * Add parentheses around `code` if in this `parent` position we need brackets
+     * @param check Position-checking function from parent
+     * @param code Code to put parentheses around
+     * @returns
      */
-    constructor(
-        private indentLevel: number = 0,
-        private readonly indentSpaces: number = 4,
-    ) {}
+    brace: (
+        position: (childPrecedence: number) => boolean,
+        code: string,
+    ) => string;
+    /**
+     * Create a position-checking function for positions where grammar rule recurses
+     * For example in `foo = "+" foo | bar` this should be used for `foo`
+     * In case of `"+" "+" bar` input, this won't put parentheses around inner
+     * `"+" bar`
+     */
+    self: (childPrecedence: number) => boolean;
+    /**
+     * Create a position-checking function for positions where grammar rule uses
+     * a higher-priority grammar rule
+     * For example in `foo = foo "-" bar | bar` this should be used for `bar`
+     * This enforces parentheses in expressions such as `x - (y - z)` where
+     * both operators have same precedence, but right-hand size is allowed to
+     * be an operator of the same precedence only if there were `()`
+     */
+    child: (childPrecedence: number) => boolean;
+};
 
-    private increaseIndent() {
-        this.indentLevel += 1;
-    }
+/**
+ * Given numeric value of precedence, where higher values stand for higher binding power,
+ * create a helper object for precedence checking
+ */
+export const makePrecedence = (myPrecedence: number): Precedence => ({
+    brace: (position, code) => (position(myPrecedence) ? `(${code})` : code),
+    self: (childPrecedence) => childPrecedence < myPrecedence,
+    child: (childPrecedence) => childPrecedence <= myPrecedence,
+});
 
-    private decreaseIndent() {
-        this.indentLevel -= 1;
-    }
+// Least binding operator
+export const lowestPrecedence = makePrecedence(0);
 
-    private indent(): string {
-        return " ".repeat(this.indentLevel * this.indentSpaces);
-    }
+export const conditionalPrecedence = makePrecedence(5);
 
-    ppAstPrimitiveTypeDecl(primitive: AstPrimitiveTypeDecl): string {
-        return `${this.indent()}primitive ${this.ppAstId(primitive.name)};`;
-    }
+export const binaryPrecedence: Readonly<
+    Record<A.AstBinaryOperation, Precedence>
+> = {
+    "||": makePrecedence(10),
```

Comment:
> since we model this part of Tact's syntax after JS

Sadly Tact's grammar is different from JS. Ternary operator and assignment are not on the same priority 2 in Tact, and priorities 15/17 are merged together.

Reply:
this is a bug and it should be fixed then

Reply:
and it wouldn't be even a violation of backwards compatibility: users should not rely on buggy behavior

Reply:
wait, assignment is not an expression in Tact

Reply:
Yes, because assignment is not an expression, it shouldn't change anything from user's perspective, but they do have different priority. Same with 15/17: we don't have a prefix `new`, so user can never distinguish whether they have different priorities anyway.